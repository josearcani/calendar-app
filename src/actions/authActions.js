import { types } from '../types/types';
import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import Swal from 'sweetalert2';

export const startLogin = (email, password) => {
  return async (dispatch) => {

    const resp = await fetchWithoutToken('auth', { email, password }, 'POST');
    const body = await resp.json(); // {name,ok,token,uid}
    
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime()); // to know when token was created

      dispatch(login({ name: body.name, uid: body.uid }));

    } else {
      Swal.fire('Error', body.msg, 'error');
    }

  }
}

export const startRegister = (name, email, password) => {
  return async (dispatch) => {

    const resp = await fetchWithoutToken('auth/new', {name, email, password }, 'POST');
    const body = await resp.json();

    // console.log('startRegister', body);
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ name: body.name, uid: body.uid }));

    } else {
      Swal.fire('Error', body.msg, 'error');
    }

  }
}

export const startChecking = () => {
  return async (dispatch) => {

    const resp = await fetchWithToken('auth/renew');
    const body = await resp.json();

    // console.log('startChecking', body);
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ name: body.name, uid: body.uid }));
      
    } else {
      dispatch(checkingFinish());
    }
  }
}

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout());
  }
}

const logout = () => ({
  type: types.authLogout
});

const login = (user) => ({
  type: types.authLogin,
  payload: user
});

const checkingFinish = () => ({
  type: types.authCheckingFinish
});
