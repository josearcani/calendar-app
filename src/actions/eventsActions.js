import { fetchWithToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepare-events';
import { types } from '../types/types';

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event
});

export const startEventAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;
    try {
      const resp = await fetchWithToken('events', event, 'POST');
      const body = await resp.json();

      if (body.ok) {
        event.id = body.event.id;
        event.user = { _id: uid, name };
        dispatch(eventAddNew(event));
      } 
    } catch (error) {
      console.log(error);
    }
  }
}

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken('events');
      const body = await resp.json(); // {ok: bool, events: arr, total: num}
      const events = prepareEvents(body.events);
      dispatch(eventLoaded(events));
    } catch (error) {
      console.log(error);
    }
  }
}

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events
})

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent,
});

export const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event
});

export const eventDeleted = () => ({
  type: types.eventDeleted,
});
