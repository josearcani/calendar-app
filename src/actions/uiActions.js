import { types } from '../types/types';

export const openModal = () => ({
  type: types.openModal,
});

export const closeModal = () => ({
  type: types.uiCloseModal,
});
