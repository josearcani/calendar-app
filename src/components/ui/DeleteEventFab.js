import React from 'react';
import { useDispatch } from 'react-redux';
import { startEventDeleted } from '../../actions/eventsActions';

export const DeleteEventFab = () => {
  const dispatch = useDispatch();

  return (
    <button
      className="btn btn-danger fab-danger"
      onClick={ () => dispatch(startEventDeleted()) }
    >
      <i className="fas fa-trash"></i>
      <span> Borrar evento</span>
    </button>
  );
};
