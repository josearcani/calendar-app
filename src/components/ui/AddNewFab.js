import React from 'react';
import { useDispatch } from 'react-redux';
import { uiOpenModal } from '../../actions/uiActions';

export const AddNewFab = () => {
  const dispatch = useDispatch();
  

  return (
    <button
      className="btn btn-primary fab"
      onClick={ () => dispatch(uiOpenModal()) }
    >
      <i className="fas fa-plus"></i>
    </button>
  );
};

