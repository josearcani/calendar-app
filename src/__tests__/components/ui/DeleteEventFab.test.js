import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { DeleteEventFab } from '../../../components/ui/DeleteEventFab';
import { startEventDeleted } from '../../../actions/eventsActions';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

// to make second test more precise
jest.mock('../../../actions/eventsActions', () => ({
  startEventDeleted: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const store = mockStore({});
store.dispatch = jest.fn(); // only need to see when dispatch something

const wrapper = mount(
  <Provider store={ store }>
    <DeleteEventFab />
  </Provider>
);

describe('Test on <DeleteEventFab />', () => {
  
  test('should <DeleteEventFab /> render normally', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  test('should call startEventDeleted when click', () => {
    wrapper.find('button').simulate('click');
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(startEventDeleted).toHaveBeenCalled();
  });
  
});
