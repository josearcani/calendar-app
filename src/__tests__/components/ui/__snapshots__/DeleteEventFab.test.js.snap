// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <DeleteEventFab /> should <DeleteEventFab /> render normally 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
     Borrar evento
  </span>
</button>
`;
